### PR TITLE
WaterTracksRenderSystem and WaterTracksObj

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/water/w3dwatertracks.cpp [hunk 1]
```diff
@@ -85,35 +392,333 @@ void WaterTracksRenderSystem::Release_Resources()
...
+        }
+
+        if (obj->m_prevSystem) {
+            obj->m_prevSystem->m_nextSystem = obj->m_nextSystem;
+        } else {
+            m_freeModules = obj->m_nextSystem;
+        }
+
+        obj->m_type = type;
+        WaterTracksObj *next = 0;
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: nullptr

File: src/game/client/water/w3dwatertracks.cpp [hunk 2]
```diff
@@ -13,15 +13,322 @@
...
+void WaterTracksObj::Init(float width, Vector2 const &start, Vector2 const &end, char *texture)
+{
+    Free_Water_Tracks_Resources();
+    m_boundingSphere.Init(Vector3(0.0f, 0.0f, 0.0f), 400.0f);
+    m_boundingBox.Init(Vector3(0.0f, 0.0f, 0.0f), Vector3(400.0f, 400.0f, 1.0f));
+    m_perpDir = end - start;
+    m_startPos = start + m_perpDir * 0.5f;
+    m_perpDir *= 1.0f / m_perpDir.Length();
+    m_waveDir = m_perpDir;
+    m_waveDir.Rotate(1.5707964f);
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: 90 deg

File: src/game/client/water/w3dwatertracks.cpp [hunk 3]
```diff
@@ -13,15 +13,322 @@
...
+    m_boundingSphere.Init(Vector3(0.0f, 0.0f, 0.0f), 400.0f);
+    m_boundingBox.Init(Vector3(0.0f, 0.0f, 0.0f), Vector3(400.0f, 400.0f, 1.0f));
+    m_x = 2;
+    m_y = 2;
+
+    m_elapsedMs = m_timeOffsetSecondWave;
+    m_startPos = start;
+    m_waveDir = end - start;
+    m_perpDir = m_waveDir;
+    m_perpDir.Rotate(-1.5707964f);
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: -90 deg

File: src/game/client/water/w3dwatertracks.cpp [hunk 4]
```diff
@@ -13,15 +13,322 @@
...
+    m_perpDir.Normalize();
+    m_waveDir = m_perpDir;
+    m_waveDir.Rotate(1.5707964f);
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: 90 deg

File: src/game/client/water/w3dwatertracks.cpp [hunk 5]
```diff
@@ -13,15 +13,322 @@
...
+
+    m_distanceFromShore = s_waveTypeInfo[m_type].distance_from_shore;
+    m_waveDir *= m_distanceFromShore;
+    m_startPos -= m_waveDir;
+    m_velocity = s_waveTypeInfo[m_type].initial_velocity;
+    m_totalMs = m_distanceFromShore / m_velocity;
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: seems to need to be  m_totalMs = (int)(m_distanceFromShore / m_velocity);

File: src/game/client/water/w3dwatertracks.cpp [hunk 6]
```diff
@@ -13,15 +13,322 @@
...
+    m_widthFraction = length * s_waveTypeInfo[m_type].initial_width_fraction;
+    m_heightFraction = m_widthFraction * s_waveTypeInfo[m_type].initial_height_fraction;
+    m_length = length;
+    m_width = width;
+
+    m_timingUnk1 = (m_distanceFromShore - m_width) / m_velocity;
+    m_velocityUnk = -(m_velocity * m_velocity) / (2.0f * m_width);
+    m_timingUnk2 = -m_velocity / m_velocityUnk;
+    m_timingUnk3 = GameMath::Sqrt(GameMath::Fabs(2.0f * m_width / m_velocityUnk));
+    m_totalMs = (m_timingUnk1 + m_timingUnk2 + m_timingUnk3);
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: appears to need to be (int)(m_timingUnk1 + m_timingUnk2 + m_timingUnk3)

File: src/game/client/water/w3dwatertracks.cpp [hunk 7]
```diff
@@ -13,15 +13,322 @@
...
+    m_scaleUnk = 2.0f * m_heightFraction / (m_timingUnk2 * m_timingUnk2);
+    m_timeToCompress = s_waveTypeInfo[m_type].time_to_compress;
+
+    if (m_type == 5) {
+        m_timingUnk3 = 1000.0f;
+        m_totalMs = (m_timingUnk1 + m_timingUnk2 + m_fadeMs + m_timingUnk3);
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: possibly missing casts?

File: src/game/client/water/w3dwatertracks.cpp [hunk 8]
```diff
@@ -13,15 +13,322 @@
...
+
+    m_timingUnk1 = (m_distanceFromShore - m_width) / m_velocity;
+    m_velocityUnk = -(m_velocity * m_velocity) / (2.0f * m_width);
+    m_timingUnk2 = -m_velocity / m_velocityUnk;
+    m_timingUnk3 = GameMath::Sqrt(GameMath::Fabs(2.0f * m_width / m_velocityUnk));
+    m_totalMs = (m_timingUnk1 + m_timingUnk2 + m_timingUnk3);
+    m_scaleUnk = 2.0f * m_heightFraction / (m_timingUnk2 * m_timingUnk2);
+    m_timeToCompress = s_waveTypeInfo[m_type].time_to_compress;
+
+    if (m_type == 5) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: enum

File: src/game/client/water/w3dwatertracks.h [hunk 9]
```diff
@@ -30,6 +33,74 @@ enum WaveType
...
+    int m_type;
+    int m_x;
+    int m_y;
+    bool m_bound;
+    Vector2 m_startPos;
+    Vector2 m_waveDir;
+    Vector2 m_perpDir;
+    Vector2 m_initialStart;
+    Vector2 m_initialEnd;
+    int m_timeOffsetSecondWave;
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.h:83]: perhaps just m_timeOffset

File: src/game/client/water/w3dwatertracks.cpp [hunk 10]
```diff
@@ -85,35 +392,333 @@ void WaterTracksRenderSystem::Release_Resources()
...
+                float r = g_theWriteableGlobalData->m_terrainAmbient[0].red;
+                float g = g_theWriteableGlobalData->m_terrainAmbient[0].green;
+                float b = g_theWriteableGlobalData->m_terrainAmbient[0].blue;
+                r = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + r;
+                g = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + g;
+                b = g_theWriteableGlobalData->m_terrainDiffuse[0].red / 2.0f + b;
+                r = r * 255.0f;
+                g = g * 255.0f;
+                b = b * 255.0f;
+                int color = (GameMath::Fast_To_Int_Truncate(r) << 16) | (GameMath::Fast_To_Int_Truncate(g) << 8)
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp:503]: color isn't used? if so add a TODO note to investigate this

File: src/game/client/water/w3dwatertracks.cpp [hunk 11]
```diff
@@ -13,15 +13,322 @@
...
+
+    m_stageZeroTexture = W3DAssetManager::Get_Instance()->Get_Texture(texture);
+}
+
+int WaterTracksObj::Render(DX8VertexBufferClass *vertex_buffer, int batch_start)
+{
+#ifdef BUILD_WITH_D3D8
+    Vector2 v1;
+    Vector2 v2;
+    float f1 = 1.0f / m_waveDir.Length();
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp]: oo_wave_dir_len

File: src/game/client/water/w3dwatertracks.cpp [hunk 12]
```diff
@@ -13,15 +13,322 @@
...
+        }
+
+        batch_start = 0;
+    }
+
+    float f2;
+    float f3 = 1.0f;
+
+    if (m_type == WAVE_TYPE_UNK) {
+        v2 = m_startPos;
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp:173]: cur_pos

File: src/game/client/water/w3dwatertracks.cpp [hunk 13]
```diff
@@ -13,15 +13,322 @@
...
+        m_elapsedMs = 0;
+        f2 = (float)m_elapsedMs / m_timingUnk1;
+        f3 = f2;
+        f3 = ((m_length - m_widthFraction) * f2 + m_widthFraction) / m_length;
+        v2 = m_startPos + (((float)m_elapsedMs * m_velocity * f1) * m_waveDir);
+        v2 -= ((m_perpDir * m_length) * 0.5f) * f3;
+        v1 = v2 - ((f1 * m_heightFraction) * m_waveDir);
+    }
+
+    Vector2 v3 = v1;
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp:248]: v3 -> vertex_pos

File: src/game/client/water/w3dwatertracks.cpp [hunk 14]
```diff
@@ -13,15 +13,322 @@
...
+#include "rtsutils.h"
+#include "shadermanager.h"
+#include "terrainlogic.h"
+#include "vertmaterial.h"
+#include "w3dshroud.h"
+#include "water.h"
 #ifdef GAME_DLL
 #include "hooker.h"
 #endif
+
```
Comment (tomsons26) [inline src/game/client/water/w3dwatertracks.cpp:35]: For VS17 which i still use to compile needs #include <cstdio> or <stdio.h> for file functions like fopen

File: src/game/client/water/w3dwatertracks.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp:503]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp:35]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp [hunk 15]
```diff
@@ -13,15 +13,324 @@
...
+            alpha = alpha / (float)m_fadeMs;
+
+            if (alpha > 1.0f) {
+                alpha = 1.0f;
+            }
+        }
+    } else if ((float)m_elapsedMs < m_timingUnk1) {
+        alpha = (float)m_elapsedMs / m_timingUnk1;
+        f3 = alpha;
+        f3 = ((m_length - m_widthFraction) * alpha + m_widthFraction) / m_length;
```
Comment (xezon) [inline src/game/client/water/w3dwatertracks.cpp]: Is this mistake? `f3` is set to alpha and the next line to something else.

File: src/game/client/water/w3dwatertracks.cpp [hunk 16]
```diff
@@ -13,15 +13,324 @@
...
+        } else {
+            v1 += (0.5f * m_scaleUnk * f4 * f4 * oo_wave_dir_len) * m_waveDir;
+        }
+
+        v1 -= ((m_perpDir * m_length) * 0.5f) * f3;
+    } else {
+        m_elapsedMs = 0;
+        alpha = (float)m_elapsedMs / m_timingUnk1;
+        f3 = alpha;
+        f3 = ((m_length - m_widthFraction) * alpha + m_widthFraction) / m_length;
```
Comment (xezon) [inline src/game/client/water/w3dwatertracks.cpp]: Is this mistake? `f3` is set to alpha and the next line to something else.

File: src/game/client/water/w3dwatertracks.cpp [hunk 17]
```diff
@@ -13,15 +13,324 @@
...
+    m_startPos -= m_waveDir;
+    m_velocity = s_waveTypeInfo[m_type].initial_velocity;
+    m_totalMs = m_distanceFromShore / m_velocity;
+    m_fadeMs = s_waveTypeInfo[m_type].time_to_fade;
+    m_widthFraction = length * s_waveTypeInfo[m_type].initial_width_fraction;
+    m_heightFraction = m_widthFraction * s_waveTypeInfo[m_type].initial_height_fraction;
+    m_length = length;
+    m_width = width;
+
+    m_timingUnk1 = (m_distanceFromShore - m_width) / m_velocity;
```
Comment (duncanspumpkin) [inline src/game/client/water/w3dwatertracks.cpp]: Time until break

File: src/game/client/water/w3dwatertracks.cpp [hunk 18]
```diff
@@ -13,15 +13,324 @@
...
+            }
+        } else if (m_timingUnk1 + m_timingUnk2 - 1000.0f < (float)m_elapsedMs) {
+            alpha = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 - 1000.0f);
+            alpha = alpha / (float)m_fadeMs;
+
+            if (alpha > 1.0f) {
+                alpha = 1.0f;
+            }
+        }
+    } else if ((float)m_elapsedMs < m_timingUnk1) {
```
Comment (duncanspumpkin) [inline src/game/client/water/w3dwatertracks.cpp]: We clearly have two different wave types here. I think structuring it as:
```
if (m_type == WAVE_TYPE_UNK){

}
else{ // All other types

}
```
would make this easier to follow.

File: src/game/client/water/w3dwatertracks.cpp [hunk 19]
```diff
@@ -13,15 +13,324 @@
...
+                * m_waveDir;
+            float f6 = (float)m_elapsedMs - (m_timingUnk1 + m_timingUnk2 + m_timeToCompress);
+            v1 += (0.5f * m_velocityUnk * f6 * f6 * oo_wave_dir_len) * m_waveDir;
+        } else {
+            v1 += (0.5f * m_scaleUnk * f4 * f4 * oo_wave_dir_len) * m_waveDir;
+        }
+
+        v1 -= ((m_perpDir * m_length) * 0.5f) * f3;
+    } else {
+        m_elapsedMs = 0;
```
Comment (duncanspumpkin) [inline src/game/client/water/w3dwatertracks.cpp:241]: This is just a repeat of (float)m_elapsedMs < m_timingUnk1 except it resets the elapsedMs first. I.e. this is restarting the loop.

File: src/game/client/water/w3dwatertracks.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

File: src/game/client/water/w3dwatertracks.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/water/w3dwatertracks.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**